analysis/preprocess/preprocess_data.R

```diff
@@ -0,0 +1,157 @@
+
+library(magrittr)
+library(dplyr)
+library(tidyverse)
+library(lubridate)
+library(data.table)
+library(readr)
+
+# Specify command arguments ----------------------------------------------------
+
+args <- commandArgs(trailingOnly=TRUE)
+print(length(args))
+if(length(args)==0){
+  cohort_name <- "vax"
+} else {
+  cohort_name <- args[[1]]
+}
+
+# Get column names -------------------------------------------------------------
+
+all_cols <- fread(paste0("output/input_",cohort_name,".csv.gz"), 
+                  header = TRUE, sep = ",", nrows = 0, 
+                  stringsAsFactors = FALSE) %>%
+  names()
+
+message("Column names found")
+
+# Identify column classes ------------------------------------------------------
+
+cat_cols <- c("patient_id", grep("_cat", all_cols, value = TRUE))
+
+bin_cols <- c(grep("_bin", all_cols, value = TRUE),
+              "has_follow_up_previous_6months", "was_alive", "has_died")
+
+num_cols <- c(grep("_num", all_cols, value = TRUE),
+              grep("vax_jcvi_age_", all_cols, value = TRUE))
+
+date_cols <- grep("_date", all_cols, value = TRUE)
+
+message("Column classes identified")
+
+# Define column classes --------------------------------------------------------
+
+col_classes <- setNames(
+  c(rep("c", length(cat_cols)),
+    rep("l", length(bin_cols)),
+    rep("d", length(num_cols)),
+    rep("D", length(date_cols))
+  ), 
+  all_cols[match(c(cat_cols, bin_cols, num_cols, date_cols), all_cols)]
+)
+
+message("Column classes defined")
+
+# Read cohort dataset ---------------------------------------------------------- 
+
+df <- read_csv(paste0("output/input_",cohort_name,".csv.gz"), 
+               col_types = col_classes)
+
+message(paste0("Dataset has been read successfully with N = ", nrow(df), " rows"))
+
+# Format columns ---------------------------------------------------------------
+
+df <- df %>%
+  mutate(across(c(contains("_date")),
+                ~ floor_date(as.Date(., format="%Y-%m-%d"), unit = "days")),
+         across(contains('_birth_year'),
+                ~ format(as.Date(., origin = "1970-01-01"), "%Y")),
+         across(contains('_num') & !contains('date'), ~ as.numeric(.)),
+         across(contains('_cat'), ~ as.factor(.)),
+         across(contains('_bin'), ~ as.logical(.)))
+
+# Overwrite vaccination information for dummy data and vax cohort only ---------
+
+if(Sys.getenv("OPENSAFELY_BACKEND") %in% c("", "expectations") &&
+   cohort_name %in% c("vax")) {
+  source("analysis/preprocess/modify_dummy_vax_data.R")
+  message("Vaccine information overwritten successfully")
+}
+
+# Describe data ----------------------------------------------------------------
+
+sink(paste0("output/describe_",cohort_name,".txt"))
+print(Hmisc::describe(df))
+sink()
+message ("Cohort ",cohort_name, " description written successfully!")
+
+# QC for consultation variable and set max to 365 (i.e., one per day) ----------
```

Comment:
We discussed moving this to the study definition. Did you make a decision about whether to do this?

Reply:
Done